src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
```

Comment:
There is no need to create a new copy of `currentState` here with `{ ...currentState }` since `currentState` is already a copy of the state and you are going to modify it directly in the subsequent lines. This is redundant and may lead to confusion.